src/codeu/chat/client/commandline/Chat.java

```diff
@@ -14,310 +14,400 @@
-  }
-
-  // Display all users known to server.
-  private void showAllUsers() {
-    clientContext.user.showAllUsers();
-  }
-
-  public boolean handleCommand(Scanner lineScanner) {
-
-    try {
-      promptForCommand();
-      doOneCommand(lineScanner);
-    } catch (Exception ex) {
-      System.out.println("ERROR: Exception during command processing. Check log for details.");
-      LOG.error(ex, "Exception during command processing");
-    }
-
-    // "alive" may have been set to false while executing a command. Return
-    // the result to signal if the user wants to keep going.
-
-    return alive;
-  }
-
-  public void selectConversation(Scanner lineScanner) {
-
-    clientContext.conversation.updateAllConversations(false);
-    final int selectionSize = clientContext.conversation.conversationsCount();
-    System.out.format("Selection contains %d entries.\n", selectionSize);
-
-    final ConversationSummary previous = clientContext.conversation.getCurrent();
-    ConversationSummary newCurrent = null;
-
-    if (selectionSize == 0) {
-      System.out.println("Nothing to select.");
-    } else {
-      final ListNavigator<ConversationSummary> navigator =
-          new ListNavigator<ConversationSummary>(
-              clientContext.conversation.getConversationSummaries(),
-              lineScanner, PAGE_SIZE);
-      if (navigator.chooseFromList()) {
-        newCurrent = navigator.getSelectedChoice();
-        clientContext.message.resetCurrent(newCurrent != previous);
-        System.out.format("OK. Conversation \"%s\" selected.\n", newCurrent.title);
-      } else {
-        System.out.println("OK. Current Conversation is unchanged.");
+    });
+
+    // M-LIST (list messages)
+    //
+    // Add a list messages command to the conversation panel. This will be called
+    // when the user enters "m-list" while on the conversation panel. This will
+    // print every message in the conversation.
+    //
+    panel.register("m-list", new Panel.Command() {
+      @Override
+      public void invoke(Scanner args) {
+        System.out.println("--- start of conversation ---");
+        for (MessageContext message = conversation.firstMessage();
+                            message != null;
+                            message = message.next()) {
+          System.out.println();
+          System.out.format("USER : %s\n", message.message.author);
+          System.out.format("SENT : %s\n", message.message.creation);
+          System.out.println();
+          System.out.println(message.message.content);
+          System.out.println();
+        }
+        System.out.println("---  end of conversation  ---");
       }
-    }
-    if (newCurrent != previous) {
-      clientContext.conversation.setCurrent(newCurrent);
-      clientContext.conversation.updateAllConversations(true);
-    }
+    });
+
+    // M-ADD (add message)
+    //
+    // Add an add message command to the conversation panel. This will be called
```

Comment:
I was trying to follow a pattern of:
  1. Adding <name> command to the <name> panel.
  2. When will the command be called (Riccardo pointed out that async code may be advance for some students).
  3. What will the command do.